### PR TITLE
Secd 550/implement nexpose client methods

### PR DESCRIPTION
**Pull request by austburn**

### CODE REVIEW

File: pkg/hydrator/nexpose.go [hunk 1]
```diff
@@ -2,38 +2,324 @@ package hydrator
...
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
 
 	"github.com/asecurityteam/nexpose-vuln-hydrator/pkg/domain"
 )
 
-type nexposeAssetVulnerability struct {
```
Comment (kconwayatlassian) [inline pkg/hydrator/nexpose.go]: question: why did you change this to an exported struct?

File: pkg/hydrator/nexpose.go [hunk 2]
```diff
@@ -2,38 +2,324 @@ package hydrator
...
+	pages := int(assetVulns.Page.TotalPages)
+	totalResources := int(assetVulns.Page.TotalResources)
+	assetVulnsChannel := make(chan NexposeAssetVulnerability, totalResources)
+	errorChan := make(chan error, pages)
+	for _, resource := range assetVulns.Resources {
+		assetVulnsChannel <- assetVulnToNexposeAssetVuln(resource)
+	}
+
+	for curPage := 1; curPage < pages; curPage = curPage + 1 {
+		go n.makePagedAssetVulnerabilitiesRequest(assetID, curPage, assetVulnsChannel, errorChan)
```
Comment (kconwayatlassian) [inline pkg/hydrator/nexpose.go]: task: While it isn't always a bad thing, passing channels into a function that writes its results to those channels should be considered a code smell. In this case you're actually doing a fan-out operation but in a slightly different way than the pattern.

I can be convinced that we don't need to use a stack of interfaces for this specific case but I'd suggest that we at least convert each function currently taking a success+error channel pair to a function that returns a value and an error. The channel management should be the responsibility of the code performing the fan-out. For example, this code would end up looking like this:

```golang
pages := int(assetVulns.Page.TotalPages)
totalResources := int(assetVulns.Page.TotalResources)
assetVulnsResults := make([] NexposeAssetVulnerability, 0, totalResources)
assetVulnsChannel := make(chan []NexposeAssetVulnerability, totalResources)
errorChan := make(chan error, pages)
for _, resource := range assetVulns.Resources {
    assetVulnsResults = append(assetVulnsResults, assetVulnToNexposeAssetVuln(resource))
}
for curPage := 1; curPage < pages; curPage = curPage + 1 {
    go func(curPage int) {
        vulns, err := n.makePagedAssetVulnerabilitiesRequest(assetID, curPage)
        if err != nil {
            errorChan <- err
            return
        }
        assetVulnsChannel <- vulns
    }(curPage)
}
// ... the rest is pretty much the same
```
There will be cases where we decide to pass channels around but they should be relatively rare or infrequent. In the vast majority of cases, our functions should do a task and optionally return a value. All concurrency related matters should be left to the code that introduces that concurrency rather than trying to negotiate concurrency from the bottom of the stack up.

File: pkg/hydrator/nexpose.go [hunk 3]
```diff
@@ -2,38 +2,324 @@ package hydrator
...
+	if err != nil {
+		errorChan <- err
+		return
+	}
+	for _, resource := range assetVulns.Resources {
+		assetVulnsChannel <- assetVulnToNexposeAssetVuln(resource)
+	}
+}
+
+func (n *NexposeClient) makeNexposeRequest(queryParams map[string]string, pathFragments ...string) ([]byte, error) {
```
Comment (kconwayatlassian) [inline pkg/hydrator/nexpose.go]: info: Something to consider that we likely don't need to action right now is: "How would you constrain the number of concurrent calls to this function?". This isn't something we typically spend a lot of time thinking about because our projects usually support a much higher concurrency and throughput than production traffic. However, this is a special case of making requests to a 3rd party system that may not be build to support the traffic we end up throwing at it.

When we talked about the layering pattern for adding concurrent fan-out we also talked through how we might limit concurrent invocations of any given function by wrapping it with a semaphore. If we use the same semaphore to wrap all of our interfaces that represent Nexpose calls then we would effectively limit the total number of concurrent calls for all APIs. However, at least one of the operations you have here is a fan-out operation to fetch the paginated vulnerability details. This means that one of the operations has a disproportionately higher cost than the other by the nature of it making N calls per invocation rather than one.

The thought experiment you might consider is what you would need to do if we decided, later, that we need to limit the total outgoing calls to Nexpose to a maximum of N concurrently.

File: pkg/hydrator/nexpose.go (hunk 1 above)
Comment (austburn) [reply pkg/hydrator/nexpose.go]: `golint` complained about returning an unexported struct

File: pkg/hydrator/nexpose.go (hunk 3 above)
Comment (austburn) [reply pkg/hydrator/nexpose.go]: Good callout. A lot of this refactoring was an attempt to reduce the burden of making a request to nexpose (auth in particular).

I'm guessing we could create an interface for the client for making the HTTP requests to Nexpose and wrap the client with rate-limiting logic. Being that the client itself implements the interface - I would think it might be possible to implement rate limiting at the `Transport` layer in order to limit the total number of concurrent requests. The interface could help inform the granularity to which you'd have the ability to control these things.
